collect_app/src/main/java/org/odk/collect/android/tasks/ChosenFileSaver.java

```diff
@@ -0,0 +1,105 @@
+package org.odk.collect.android.tasks;
+
+import android.content.Context;
+import android.net.Uri;
+import android.provider.MediaStore;
+
+import org.odk.collect.android.R;
+import org.odk.collect.android.application.Collect;
+import org.odk.collect.android.dao.helpers.ContentResolverHelper;
+import org.odk.collect.android.exception.GDriveConnectionException;
+import org.odk.collect.android.logic.FormController;
+import org.odk.collect.android.utilities.FileUtils;
+import org.odk.collect.android.utilities.MediaUtils;
+
+import java.io.File;
+import java.lang.ref.WeakReference;
+
+import io.reactivex.Observable;
+import timber.log.Timber;
+
+public class ChosenFileSaver {
+
+    private final WeakReference weakReferenceContext;
+    private final Uri selectedFile;
+    private File newFile;
+    private int errorMessageRes;
+
+    private ChosenFileSaver(Context context, Uri selectedFile) {
+        weakReferenceContext = new WeakReference<>(context);
+        this.selectedFile = selectedFile;
+    }
+
+    public static Observable<SaveResult> createObservable(Context context, Uri selectedFile) {
+        return Observable.create(emitter -> {
+            ChosenFileSaver fileSaver = new ChosenFileSaver(context, selectedFile);
+            boolean result = fileSaver.saveChosenFile();
+
+            SaveResult saveResult = new SaveResult(result, fileSaver.getSavedFile(), fileSaver.getErrorMessageRes());
+            emitter.onNext(saveResult);
+            emitter.onComplete();
+        });
+    }
+
+    private int getErrorMessageRes() {
+        return errorMessageRes;
+    }
+
+    private File getSavedFile() {
+        return newFile;
+    }
+
+    private boolean saveChosenFile() {
+        Context context = (Context) weakReferenceContext.get();
+        FormController formController = Collect.getInstance().getFormController();
+        try {
+            if (context != null && formController != null && formController.getInstanceFile() != null) {
```

Comment:
I took these checks out of my code because the exceptions would be handled via `onError` but it's something I would revisit because some null checks are good and can be handled more gracefully.

Reply:
I feel like adding more null checks make the code more predictable and easier to anticipate

Reply:
But, preserving stack trace is what will be better than logging custom messages. You are right